README.md
ownCloud Cookbook
=================
[![Cookbook Version](https://img.shields.io/cookbook/v/owncloud.svg?style=flat)](https://supermarket.chef.io/cookbooks/owncloud)
[![GitHub Source](https://img.shields.io/badge/source-GitHub-blue.svg?style=flat)](https://github.com/zuazo/owncloud-cookbook)
[![Dependency Status](http://img.shields.io/gemnasium/zuazo/owncloud-cookbook.svg?style=flat)](https://gemnasium.com/zuazo/owncloud-cookbook)
[![Code Climate](http://img.shields.io/codeclimate/github/zuazo/owncloud-cookbook.svg?style=flat)](https://codeclimate.com/github/zuazo/owncloud-cookbook)
[![Build Status](http://img.shields.io/travis/zuazo/owncloud-cookbook.svg?style=flat)](https://travis-ci.org/zuazo/owncloud-cookbook)
[![Coverage Status](http://img.shields.io/coveralls/zuazo/owncloud-cookbook.svg?style=flat)](https://coveralls.io/r/zuazo/owncloud-cookbook?branch=master)
[![Inline docs](http://inch-ci.org/github/zuazo/owncloud-cookbook.svg?branch=master&style=flat)](http://inch-ci.org/github/zuazo/owncloud-cookbook)

[Chef](https://www.chef.io/) cookbook to install and configure [ownCloud](http://owncloud.org/), an open source personal cloud for data and file sync, share and view.

Requirements
============

## Supported Platforms

* CentOS
* Debian
* Scientific Linux
* Ubuntu

The following platforms are tested with Test Kitchen:

| Platform           | MySQL      | PostgreSQL | SQLite     | NGINX      |
|:-------------------|:----------:|:----------:|:----------:|:-----------|
| CentOS 5           | ✔          |            |            |            |
| CentOS 6           | ✔ (travis) |            | ✔ (travis) |            |
| CentOS 7           | ✔          | ✔          | ✔          | ✔          |
| Debian 6           | ✔          |            | ✔          | ✔          |
| Debian 7           | ✔ (travis) | ✔ (travis) | ✔ (travis) | ✔ (travis) |
| Debian 8           | ✔ (travis) | ✔ (travis) | ✔ (travis) | ✔ (travis) |
| Ubuntu 12.04       | ✔          | ✔ (travis) | ✔ (travis) | ✔          |
| Ubuntu 14.04       | ✔          | ✔ (travis) | ✔ (travis) | ✔          |
| Ubuntu 15.04       | ✔          |            | ✔ (travis) | ✔          |
| Scientific Linux 6 | ✔ (travis) |            | ✔ (travis) |            |

Please, [let me know](https://github.com/zuazo/owncloud-cookbook/issues/new?title=I%20have%20used%20it%20successfully%20on%20...) if you use it successfully on any other platform.

## Required Cookbooks

* [apache2](https://supermarket.chef.io/cookbooks/apache2/)
* [apt](https://supermarket.chef.io/cookbooks/apt/)
* [cron](https://supermarket.chef.io/cookbooks/cron/)
* [database](https://supermarket.chef.io/cookbooks/database/)
* [encrypted_attributes](https://supermarket.chef.io/cookbooks/encrypted_attributes/)
* [mysql2_chef_gem](https://supermarket.chef.io/cookbooks/mysql2_chef_gem/)
* [mysql](https://supermarket.chef.io/cookbooks/mysql/)
* [nginx](https://supermarket.chef.io/cookbooks/nginx/)
* [openssl](https://supermarket.chef.io/cookbooks/openssl/)
* [php](https://supermarket.chef.io/cookbooks/php/)
* [php-fpm](https://supermarket.chef.io/cookbooks/php-fpm/)
* [postfix](https://supermarket.chef.io/cookbooks/postfix/)
* [postgresql](https://supermarket.chef.io/cookbooks/postgresql/)
* [ssl_certificate](https://supermarket.chef.io/cookbooks/ssl_certificate/)

## Required Applications

* Chef `>= 11.14.2`.
* Ruby `2.0.0` or higher.

## Other Requirements

On RedHat based platforms, you need to disable or configure SELinux correctly. You can use the `selinux::disabled` recipe for that.

Attributes
==========

| Attribute                                           | Default       | Description                    |
|:----------------------------------------------------|:--------------|:-------------------------------|
| `node['owncloud']['version']`                       | *calculated*  | Version of ownCloud to install.
| `node['owncloud']['download_url']`                  | *calculated*  | URL from where ownCloud will be downloaded.
| `node['owncloud']['deploy_from_git']`               | `false`       | Whether ownCloud should be deployed from the git repository.
| `node['owncloud']['git_repo']`                      | *calculated*  | URL of the ownCloud git repository.
| `node['owncloud']['git_ref']`                       | `nil`         | Git reference to deploy.
| `node['owncloud']['www_dir']`                       | *calculated*  | Root directory defined in the web server where web documents are stored.
| `node['owncloud']['dir']`                           | *calculated*  | Directory where ownCloud will be installed.
| `node['owncloud']['data_dir']`                      | *calculated*  | Directory where ownCloud data will be stored.
| `node['owncloud']['server_name']`                   | *calculated*  | Sets the server name for the ownCloud virtual host.
| `node['owncloud']['server_aliases']`                | `[]`          | Sets the server name aliases for the ownCloud virtual host.
| `node['owncloud']['install_postfix']`               | `true`        | Whether to install Postfix when a local MTA is needed.
| `node['owncloud']['web_server']`                    | `'apache'`    | Web server to use: `'apache'` or `'nginx'`.
| `node['owncloud']['php-fpm']['pool']`               | `'owncloud'`  | PHP-FPM pool name to use with ownCloud.
| `node['owncloud']['max_upload_size']`               | `'512M'`      | Maximum allowed size for uploaded files.
| `node['owncloud']['sendfile']`                      | *calculated*  | Whether to enable *sendfile* on web server. You should set to `'false'` if you use NFS or SMB mounts.
| `node['owncloud']['ssl']`                           | `true`        | Whether ownCloud should accept requests through TLS.
| `node['owncloud']['admin']['user']`                 | `'admin'`     | Administrator username.
| `node['owncloud']['admin']['pass']`                 | *calculated*  | Administrator password.
| `node['owncloud']['config']['dbtype']`              | `'mysql'`     | Type of database, can be `'sqlite'`, `'mysql'` or `'pgsql'`.
| `node['owncloud']['config']['dbname']`              | `'owncloud'`  | Name of the ownCloud database.
| `node['owncloud']['config']['dbuser']`              | `'owncloud'`  | User to access the ownCloud database.
| `node['owncloud']['config']['dbpassword']`          | *calculated*  | Password to access the ownCloud database.
| `node['owncloud']['config']['dbhost']`              | `'127.0.0.1'` | Host running the ownCloud database.
| `node['owncloud']['config']['dbport']`              | *calculated*  | Port the ownCloud database is running at.
| `node['owncloud']['config']['dbtableprefix']`       | `''`          | Prefix for the ownCloud tables in the database.
| `node['owncloud']['config']['mail_smtpmode']`       | `'sendmail'`  | Mode to use for sending mail, can be `'sendmail'`, `'smtp'`, `'qmail'` or `'php'`.
| `node['owncloud']['config']['mail_smtphost']`       | `'127.0.0.1'` | Host to use for sending mail, depends on *mail_smtpmode* if this is used.
| `node['owncloud']['config']['mail_smtpport']`       | `25`          | Port to use for sending mail, depends on *mail_smtpmode* if this is used.
| `node['owncloud']['config']['mail_smtptimeout']`    | `10`          | SMTP server timeout in seconds for sending mail, depends on *mail_smtpmode* if this is used.
| `node['owncloud']['config']['mail_smtpsecure']`     | `''`          | SMTP connection prefix or sending mail, depends on *mail_smtpmode* if this is used. Can be `''`, `'ssl'` or `'tls'`.
| `node['owncloud']['config']['mail_smtpauth']`       | `false`       | Whether authentication is needed to send mail, depends on *mail_smtpmode* if this is used.
| `node['owncloud']['config']['mail_smtpauthtype']`   | `'LOGIN'`     | Authentication type needed to send mail, depends on *mail_smtpmode* if this is used. Can be `'LOGIN'`, `'PLAIN'` or `'NTLM'`.
| `node['owncloud']['config']['mail_smtpname']`       | `''`          | Username to use for sendmail mail, depends on *mail_smtpauth* if this is used.
| `node['owncloud']['config']['mail_smtppassword']`   | `''`          | Password to use for sendmail mail, depends on *mail_smtpauth* if this is used.
| `node['owncloud']['cron']['enabled']`               | `true`        | Whether to enable ownCloud cron job.
| `node['owncloud']['cron']['min']`                   | `'*/15'`      | Minute to run ownCloud cron at.
| `node['owncloud']['cron']['hour']`                  | `'*'`         | Hour to run ownCloud cron at.
| `node['owncloud']['cron']['day']`                   | `'*'`         | Day of month to run ownCloud cron at.
| `node['owncloud']['cron']['month']`                 | `'*'`         | Month to run ownCloud cron at.
| `node['owncloud']['cron']['weekday']`               | `'*'`         | Weekday to run ownCloud cron at.
| `node['owncloud']['skip_permissions']`              | `false`       | Whether to skip settings the permissions of the ownCloud directory. Set this to `true` when using NFS synced folders.
| `node['owncloud']['manage_database']`               | *calculated*  | Whether to manage database creation.
| `node['owncloud']['mysql']['instance']`             | `'default'`   | MySQL database instance name to run by the `mysql_service` LWRP from the [mysql](https://supermarket.chef.io/cookbooks/mysql) cookbook.
| `node['owncloud']['mysql']['data_dir']`             | *calculated*  | MySQL data files path.
| `node['owncloud']['mysql']['run_group']`            | *calculated*  | MySQL system group.
| `node['owncloud']['mysql']['run_user']`             | *calculated*  | MySQL system user.
| `node['owncloud']['mysql']['version']`              | `nil`         | MySQL version to install by the `mysql_service` LWRP. Refer to [`mysql` cookbook platform support section](https://github.com/chef-cookbooks/mysql#platform-support).
| `node['owncloud']['mysql']['server_root_password']` | *calculated*  | MySQL root password to access a database instance.
| `node['owncloud']['encrypt_attributes']`            | `false`       | Whether to encrypt ownCloud attributes containing credential secrets.

## Platform Support Related Attributes

Some cookbook attributes are used internally to support the different platforms. Surely you want to change them if you want to support new platforms or want to improve the support of some platforms already supported.

| Attribute                                | Default       | Description                       |
|:-----------------------------------------|:--------------|:----------------------------------|
| `node['owncloud']['packages']['core']`   | *calculated*  | ownCloud core package names as array.
| `node['owncloud']['packages']['sqlite']` | *calculated*  | ownCloud package names array for SQLite.
| `node['owncloud']['packages']['mysql']`  | *calculated*  | ownCloud package names array for MySQL.
| `node['owncloud']['packages']['pgsql']`  | *calculated*  | ownCloud package names array for PostgreSQL.

Recipes
=======

## owncloud::default

Installs and configures ownCloud.

Usage
=====

Add `recipe[owncloud]` to your node's run list or role, or include it in another cookbook.

The back-end database will be [MySQL](http://www.mysql.com/) by default, but [PostgreSQL](http://www.postgresql.org/) and [SQLite](http://www.sqlite.org/) databases can also be used. Database type can be set on `node['owncloud']['config']['dbtype']`, supported values are `mysql`, `pgsql` and `sqlite`.

On the first run, several passwords will be automatically generated and stored in the node:

* `node['owncloud']['admin']['pass']`
* `node['owncloud']['config']['dbpassword']` (Only when using *MySQL* or *PostgreSQL*)
* `node['owncloud']['mysql']['server_root_password']` (Only when using *MySQL*)

When using Chef Solo, these passwords need to be set manually.

The attribute `node['owncloud']['server_name']` should be set to the domain name used by the ownCloud installation. If not set, `node['fqdn']` will be used instead.

By default ownCloud cookbook relies on a local *Postfix* installation to send emails. But a remote SMTP server can be used changing `node['owncloud']['config']['mail_smtpmode']` to `smtp` and setting up the rest of the mail configuration attributes (see example below).

## Examples

### Basic ownCloud Role

```ruby
name 'owncloud'
description 'Install ownCloud'
default_attributes(
  'owncloud' => {
    'server_name' => 'cloud.mysite.com'
  }
)
run_list(
  'recipe[owncloud]'
)
```

### Using Remote SMTP Server

In this example an [Amazon Simple Email Service](http://aws.amazon.com/ses/) account is used to send emails.

```ruby
name 'owncloud_ses'
description 'Install ownCloud and use an AWS SES account to send emails'
default_attributes(
  'owncloud' => {
    'server_name' => 'cloud.mysite.com',
    'config' => {
      'mail_smtpmode' => 'smtp',
      'mail_smtphost' => 'email-smtp.us-east-1.amazonaws.com',
      'mail_smtpport' => 465,
      'mail_smtpsecure' => 'tls',
      'mail_smtpauth' => true,
      'mail_smtpauthtype' => 'LOGIN',
      'mail_smtpname' => 'amazon-ses-username',
      'mail_smtppassword' => 'amazon-ses-password'
    }
  }
)
run_list(
  'recipe[owncloud]'
)
```

### Deploying from Git

The ownCloud code can be deployed from the Git repository. Git recipe must be included on the run_list.

```ruby
name 'owncloud_git'
description 'Install ownCloud from Git'
default_attributes(
  'owncloud' => {
    'server_name' => 'cloud.mysite.com',
    'deploy_from_git' => true,
    'git_ref' => 'master'
  }
)
run_list(
  'recipe[git::default]',
  'recipe[owncloud]'
)
```

## The HTTPS Certificate

ownCloud will accept HTTPS requests when `node['owncloud']['ssl']` is set to `true`. By default the cookbook will create a self-signed certificate, but a custom one can also be used.

The custom certificate can be read from several sources:

* Attribute
* Data Bag
* Encrypted Data Bag
* Chef Vault
* File

This cookbook uses the [`ssl_certificate`](https://supermarket.chef.io/cookbooks/ssl_certificate) cookbook to create the HTTPS certificate. The namespace used is `node['owncloud']`. For example:

```ruby

CHANGELOG.md
# CHANGELOG for owncloud

This file is used to list changes made in each version of `owncloud`.

## v1.0.0 (2015-10-08)

### Upgrading from a `0.x.y` Cookbook Release

***Note:*** Please do this with caution. Make a full backup before upgrading.

If you want to upgrade the cookbook version from a `0.x` release, you should change the MySQL data directory to the old path (or migrate the database by hand):

```ruby
node.default['owncloud']['mysql']['data_dir'] = '/var/lib/mysql'
# [...]
include_recipe 'owncloud'
```

### Breaking Changes on v1.0.0

* Use [`mysql`](https://supermarket.chef.io/cookbooks/mysql) cookbook version `6` ([issue #21](https://github.com/zuazo/owncloud-cookbook/pull/21), thanks [@avsh](https://github.com/avsh)) ([#bd4eba9](https://github.com/zuazo/owncloud-cookbook/commit/bd4eba9d26186baa3d3033b52449aedbf62e4bf8), [#22a2a4f](https://github.com/zuazo/owncloud-cookbook/commit/22a2a4f3fa9d6bd26598324298e6e9effa086b8a), [#f686605](https://github.com/zuazo/owncloud-cookbook/commit/f686605339db550a249bd79f8a8c306c9938617d), [#33a393c](https://github.com/zuazo/owncloud-cookbook/commit/33a393cf871c489d8282338e4fe8e0e71da9760c), [#f7d467d](https://github.com/zuazo/owncloud-cookbook/commit/f7d467db71082f4e4f465bac271f2f3434815052), [#0e63402](https://github.com/zuazo/owncloud-cookbook/commit/0e63402b8a059c449d67591a43ae26e832088bb9), [#5f04c62](https://github.com/zuazo/owncloud-cookbook/commit/5f04c622da3990cb25c04c7666fd9961d90ce21d), [#778211d](https://github.com/zuazo/owncloud-cookbook/commit/778211d22afb9c8a56123ae86438ddb4903b0d68)).
* Requires **Chef `11.14.2`**.
* Drop Ruby `1.9.3` support (required by foodcritic `5`): **Ruby `2` required** ([#952caa4](https://github.com/zuazo/owncloud-cookbook/commit/952caa46506db776a44bf49465058f9454f17345)).
* Drop Ubuntu `10.04` support ([EOL reached](http://fridge.ubuntu.com/2015/04/30/ubuntu-10-04-lucid-lynx-end-of-life-reached-on-april-30-2015/)) ([#a7ad0c3](https://github.com/zuazo/owncloud-cookbook/commit/a7ad0c383cd8f3dc3970d1aca6415fc4298f9fdb)).
* Improve some old platforms support (**installs old ownCloud versions on those platforms**) ([#b3dc45e](https://github.com/zuazo/owncloud-cookbook/commit/b3dc45ec6174efb76b8ae5b6e13a11804180b18a), [#afeb353](https://github.com/zuazo/owncloud-cookbook/commit/afeb35334323024077444bdf72bd8f913bdb9e29), [#cf48390](https://github.com/zuazo/owncloud-cookbook/commit/cf48390e15f503078d66b005d0573d2e899b4a10), [#336c270](https://github.com/zuazo/owncloud-cookbook/commit/336c2706d065abf1c22989e8944e5c19df5f06b4)):
  * Debian `6`.
  * CentOS `6`.

### New Features on v1.0.0

* Add Ubuntu `15.04` support ([#eb328b6](https://github.com/zuazo/owncloud-cookbook/commit/eb328b6b50cf8083d93225e05aa8fac8ed7e58fe), [#2d42783](https://github.com/zuazo/owncloud-cookbook/commit/2d427837099db02bd332b2bbb5ad5d15525a5e3e), [#d43a45b](https://github.com/zuazo/owncloud-cookbook/commit/d43a45b2c120c3f3e764ae40e7ad390b0dc9f5bc)).
* Add Scientific Linux support ([#f91ce0a](https://github.com/zuazo/owncloud-cookbook/commit/f91ce0a8d4cb3dd612b170d111e0002a1b1bfb55)).
* Support non standard database ports ([issue #22](https://github.com/zuazo/owncloud-cookbook/pull/22), thanks [Michael Sprauer](https://github.com/MichaelSp)) ([#7cdb4be](https://github.com/zuazo/owncloud-cookbook/commit/7cdb4beb07f73ace9936776a1845e4942071c13d), [#772844c](https://github.com/zuazo/owncloud-cookbook/commit/772844cf2180ae66f359bc03875642c666c5196e), [#7f9ce9c](https://github.com/zuazo/owncloud-cookbook/commit/7f9ce9c59f152c6e5f0d134d5570e7e1c12bd3d2), [#661b706](https://github.com/zuazo/owncloud-cookbook/commit/661b7062cdbac2de151cd6fb19cf2193ef372a39), [#170358e](https://github.com/zuazo/owncloud-cookbook/commit/170358ef0b0e322e69e9594ed59605acb0d1e2c5)).
* Use [`ssl_certificate`](https://supermarket.chef.io/cookbooks/ssl_certificate) cookbook for the HTTPS certificate (requires Chef `11.14.2`, apart from that, it should be backward compatible) ([#f08b336](https://github.com/zuazo/owncloud-cookbook/commit/f08b33620cc3f994a0d3735c44fc573753eb02ba), [#ed8ec74](https://github.com/zuazo/owncloud-cookbook/commit/ed8ec744aa077fc7d2267acbce11e0d826659cb6), [#214c57b](https://github.com/zuazo/owncloud-cookbook/commit/214c57b0c7b4b8a1100689622bebd8a029a0559c), [#65ebf38](https://github.com/zuazo/owncloud-cookbook/commit/65ebf38f00b72572140a16f384c357ed59f1808e)).
* Generate download URL dynamically using the version value ([#553307a](https://github.com/zuazo/owncloud-cookbook/commit/553307ad4310bfd7d4c9f5565fa07dc76faa111d)).
* Add `node['owncloud']['manage_database']` attribute ([#7fd36c7](https://github.com/zuazo/owncloud-cookbook/commit/7fd36c7ee62cd477735d82dc5a0aea69d4d1290c)).
* Encrypt some attributes with [`chef-encrypted-attributes`](http://onddo.github.io/chef-encrypted-attributes/) gem ([#acb190e](https://github.com/zuazo/owncloud-cookbook/commit/acb190eedfa1d577972f395421b5fa4fc5f14c9e)).
* Move packages to install to node attributes ([#17af639](https://github.com/zuazo/owncloud-cookbook/commit/17af6394a5f3780a9c82ef59496fd09be4f915a9)).
* Add a Dockerfile ([#489e379](https://github.com/zuazo/owncloud-cookbook/commit/489e379d0ae652ae1fb6ba70e52388ba7758a503)).

### Fixes on v1.0.0

* Install crontab: Fixes a CentOS error ([#7181d5a](https://github.com/zuazo/owncloud-cookbook/commit/7181d5a8a5d9f52b04ed39e509176e82fb3d5eb9)).
* `OwncloudCookbook::Config` library: Escape JSON backslash and quotes for PHP ([#4941aea](https://github.com/zuazo/owncloud-cookbook/commit/4941aea170398190e83cdd103d7a012eb335c341)).
* Fix `postgresql_database_user[owncloud]` resource duplication ([#70e48e6](https://github.com/zuazo/owncloud-cookbook/commit/70e48e691d207cd575f32369579064deb917fa10)).

### Improvements on v1.0.0

* Remove `#deep_to_hash` method (requires Chef `11.12`) ([#cadff9e](https://github.com/zuazo/owncloud-cookbook/commit/cadff9e7d193a8a47f1c560df469605ba043a154)).

metadata.rb
# encoding: UTF-8
#
# Cookbook Name:: owncloud
# Author:: Xabier de Zuazo (<xabier@zuazo.org>)
# Copyright:: Copyright (c) 2015 Xabier de Zuazo
# Copyright:: Copyright (c) 2013-2015 Onddo Labs, SL.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

name 'owncloud'
maintainer 'Xabier de Zuazo'
maintainer_email 'xabier@zuazo.org'
license 'Apache 2.0'
description <<-EOH
Installs and configures ownCloud, an open source personal cloud for data and
file sync, share and view.
EOH
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version '1.1.0' # WiP

if respond_to?(:source_url)
  source_url "https://github.com/zuazo/#{name}-cookbook"
end
if respond_to?(:issues_url)
  issues_url "https://github.com/zuazo/#{name}-cookbook/issues"
end

supports 'centos'
supports 'debian'
supports 'scientific'
supports 'ubuntu'

depends 'apache2', '~> 3.0'
depends 'apt', '~> 2.0'
depends 'cron', '~> 1.6'
depends 'database', '~> 4.0'
depends 'encrypted_attributes', '~> 0.2'
depends 'mysql2_chef_gem', '~> 1.0.1'
depends 'mysql', '~> 6.0'
depends 'nginx', '~> 2.7'
depends 'openssl', '~> 4.0'
depends 'php', '~> 1.4'
depends 'php-fpm', '~> 0.7'
depends 'postfix', '~> 3.0'
depends 'postgresql', '~> 3.4.18'
depends 'ssl_certificate', '~> 1.1'

recipe 'owncloud::default', 'Installs and configures ownCloud'

attribute 'owncloud/version',
          display_name: 'ownCloud Version',
          description: 'Version of ownCloud to install',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/download_url',
          display_name: 'ownCloud Download Url',
          description: 'Url from where ownCloud will be downloaded',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/deploy_from_git',
          display_name: 'ownCloud Deploy From Git',
          description:
            'Whether ownCloud should be deployed from the git repository',
          type: 'string',
          required: 'optional',
          default: 'false'

attribute 'owncloud/git_repo',
          display_name: 'ownCloud Git Repo',
          description: 'Url of the ownCloud git repository',
          type: 'string',
          required: 'optional',
          default: 'https://github.com/owncloud/core.git'

attribute 'owncloud/git_ref',
          display_name: 'ownCloud Git Ref',
          description: 'Git reference to deploy',
          type: 'string',
          required: 'optional',
          default: 'nil'

attribute 'owncloud/www_dir',
          display_name: 'ownCloud www Dir',
          description:
            'Root directory defined in Apache where web documents are stored',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/dir',
          display_name: 'ownCloud Dir',
          description: 'Directory where ownCloud will be installed',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/data_dir',
          display_name: 'ownCloud Data Dir',
          description: 'Directory where ownCloud data will be stored',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/server_name',
          display_name: 'ownCloud Server Name',
          description: 'Sets the server name for the ownCloud virtual host',
          calculated: true,
          type: 'string',
          required: 'recommended'

attribute 'owncloud/server_aliases',
          display_name: 'ownCloud Server Aliases',
          description:
            'Sets the server name aliases for the ownCloud virtual host',
          type: 'array',
          required: 'optional',
          default: []

attribute 'owncloud/install_postfix',
          display_name: 'install Postfix?',
          description: 'Whether to install Postfix when a local MTA is needed',
          choice: %w(true false),
          type: 'string',
          required: 'optional',
          default: 'true'

attribute 'owncloud/web_server',
          display_name: 'Web Server',
          description: 'Web server to use: apache or nginx',
          choice: %w(apache nginx),
          type: 'string',
          required: 'optional',
          default: 'apache'

attribute 'owncloud/php-fpm/pool',
          display_name: 'PHP-FPM pool',
          description: 'PHP-FPM pool name to use with ownCloud',
          type: 'string',
          required: 'optional',
          default: 'owncloud'

attribute 'owncloud/max_upload_size',
          display_name: 'Max Upload Size',
          description: 'Maximum allowed size for uploaded files',
          type: 'string',
          required: 'optional',
          default: '512M'

attribute 'owncloud/sendfile',
          display_name: 'Sendfile',
          description:
            'Whether to enable Sendfile on web server. You should set to '\
            'false if you use NFS or SMB mounts',
          choice: %w(true false),
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/ssl',
          display_name: 'ownCloud Use SSL?',
          description: 'Whether ownCloud should accept requests through SSL',
          choice: %w(true false),
          type: 'string',
          required: 'optional',
          default: 'true'

attribute 'owncloud/admin/user',
          display_name: 'ownCloud Admin Username',
          description: 'Administrator username',
          type: 'string',
          required: 'optional',
          default: 'admin'

attribute 'owncloud/admin/pass',
          display_name: 'ownCloud Admin Password',
          description: 'Administrator password',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/config/dbtype',
          display_name: 'ownCloud Database Type',
          description: 'Type of database, can be sqlite, mysql or pgsql',
          choice: %w(mysql pgsql sqlite),
          type: 'string',
          required: 'optional',
          default: 'mysql'

attribute 'owncloud/config/dbname',
          display_name: 'ownCloud Database Name',
          description: 'Name of the ownCloud database',
          type: 'string',
          required: 'optional',
          default: 'owncloud'

attribute 'owncloud/config/dbuser',
          display_name: 'ownCloud Database User',
          description: 'User to access the ownCloud database',
          type: 'string',
          required: 'optional',
          default: 'owncloud'

attribute 'owncloud/config/dbpassword',
          display_name: 'ownCloud Database Password',
          description: 'Password to access the ownCloud database',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/config/dbhost',
          display_name: 'ownCloud Database Host',
          description: 'Host running the ownCloud database',
          type: 'string',
          required: 'optional',
          default: '127.0.0.1'

attribute 'owncloud/config/dbport',
          display_name: 'ownCloud Database Port',
          description: 'Port the ownCloud database is running at',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/config/dbtableprefix',
          display_name: 'ownCloud Database Table Prefix',
          description: 'Prefix for the ownCloud tables in the database',
          type: 'string',
          required: 'optional',
          default: ''

attribute 'owncloud/config/mail_smtpmode',
          display_name: 'ownCloud Mail SMTP Mode',
          description:
            'Mode to use for sending mail, can be sendmail, smtp, qmail or php',
          choice: %w(sendmail smtp qmail php),
          type: 'string',
          required: 'optional',
          default: 'sendmail'

attribute 'owncloud/config/mail_smtphost',
          display_name: 'ownCloud Mail SMTP Host',
          description:
            'Host to use for sending mail, depends on mail_smtpmode if this '\
            'is used',
          type: 'string',
          required: 'optional',
          default: '127.0.0.1'

attribute 'owncloud/config/mail_smtpport',
          display_name: 'ownCloud Mail SMTP Port',
          description:
            'Port to use for sending mail, depends on mail_smtpmode if this '\
            'is used',
          type: 'string',
          required: 'optional',
          default: '25'

attribute 'owncloud/config/mail_smtptimeout',
          display_name: 'ownCloud Mail SMTP Timeout',
          description:
            'SMTP server timeout in seconds for sending mail, depends on '\
            'mail_smtpmode if this is used',
          type: 'string',
          required: 'optional',
          default: '10'

attribute 'owncloud/config/mail_smtpsecure',
          display_name: 'ownCloud Mail SMTP Secure',
          description:
            'SMTP connection prefix or sending mail, depends on mail_smtpmode '\
            'if this is used. Can be "", ssl or tls',
          choice: ['', 'ssl', 'tls'],
          type: 'string',
          required: 'optional',
          default: ''

attribute 'owncloud/config/mail_smtpauth',
          display_name: 'ownCloud Mail SMTP Auth',
          description:
            'Whether authentication is needed to send mail, depends on '\
            'mail_smtpmode if this is used',
          choice: %w(true false),
          type: 'string',
          required: 'optional',
          default: 'false'

attribute 'owncloud/config/mail_smtpauthtype',
          display_name: 'ownCloud Mail SMTP Auth Type',
          description:
            'Authentication type needed to send mail, depends on '\
            'mail_smtpmode if this is used. Can be LOGIN, PLAIN or NTLM',
          choice: %w(LOGIN PLAIN NTLM),
          type: 'string',
          required: 'optional',
          default: 'LOGIN'

attribute 'owncloud/config/mail_smtpname',
          display_name: 'ownCloud Mail SMTP Name',
          description:
            'Username to use for sendmail mail, depends on mail_smtpauth if '\
            'this is used',
          type: 'string',
          required: 'optional',
          default: ''

attribute 'owncloud/config/mail_smtppassword',
          display_name: 'ownCloud Mail SMTP Password',
          description:
            'Password to use for sendmail mail, depends on mail_smtpauth if '\
            'this is used',
          type: 'string',
          required: 'optional',
          default: ''

attribute 'owncloud/cron/enabled',
          display_name: 'ownCloud cron enabled',
          description: 'Whether to enable ownCloud cron',
          type: 'string',
          required: 'optional',
          default: 'true'

attribute 'owncloud/cron/min',
          display_name: 'ownCloud cron minute',
          description: 'Minute to run ownCloud cron at',
          type: 'string',
          required: 'optional',
          default: '*/15'

attribute 'owncloud/cron/hour',
          display_name: 'ownCloud cron hour',
          description: 'Hour to run ownCloud cron at',
          type: 'string',
          required: 'optional',
          default: '*'

attribute 'owncloud/cron/day',
          display_name: 'ownCloud cron day',
          description: 'Day of month to run ownCloud cron at',
          type: 'string',
          required: 'optional',
          default: '*'

attribute 'owncloud/cron/month',
          display_name: 'ownCloud cron month',
          description: 'Month to run ownCloud cron at',
          type: 'string',
          required: 'optional',
          default: '*'

attribute 'owncloud/cron/weekday',
          display_name: 'ownCloud cron weekday',
          description: 'Weekday to run ownCloud cron at',
          type: 'string',
          required: 'optional',
          default: '*'

attribute 'owncloud/skip_permissions',
          display_name: 'Skip permissions',
          description:
            'Whether to skip settings the permissions of the ownCloud '\
            'directory. Set this to true when using NFS synced folders.',
          choice: %w(true false),
          type: 'string',
          required: 'optional'

attribute 'owncloud/manage_database',
          display_name: 'database manage',
          description: 'Whether to manage database creation.',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/mysql/instance',
          display_name: 'MySQL database instance name',
          description:
            'MySQL database instance name to run by the mysql_service lwrp '\
            'from the mysql cookbook.',
          type: 'string',
          required: 'optional',
          default: 'default'

attribute 'owncloud/database/data_dir',
          display_name: 'MySQL server data dir',
          description: 'MySQL data files path.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'owncloud/database/run_group',
          display_name: 'MySQL run group',
          description: 'MySQL system group.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'owncloud/database/run_user',
          display_name: 'MySQL run user',
          description: 'MySQL system user.',
          type: 'string',
          required: 'optional',
          calculated: true

attribute 'owncloud/database/version',
          display_name: 'MySQL server version',
          description:
            'MySQL version to install by the mysql_service lwrp. Refer to '\
            'https://github.com/chef-cookbooks/mysql#platform-support',
          type: 'string',
          required: 'optional',
          default: 'nil'

attribute 'owncloud/mysql/server_root_password',
          display_name: 'Database Root Password',
          description: 'Database admin password to access a database instance.',
          calculated: true,
          type: 'string',
          required: 'optional'

attribute 'owncloud/encrypt_attributes',
          display_name: 'owncloud encrypt attributes',
          description: 'Whether to encrypt ownCloud attributes containing '\
            'credential secrets.',
          type: 'string',
          choice: %w(true false),
          default: 'false'

grouping 'owncloud/packages',
         title: 'owncloud packages',
         description: 'ownCloud packages'

attribute 'owncloud/packages/core',
          display_name: 'owncloud packages core',
          description: 'ownCloud core package names as array.',
          type: 'array',
          required: 'optional',
          calculated: true

attribute 'owncloud/packages/sqlite',
          display_name: 'owncloud packages sqlite',
          description: 'ownCloud package names array for SQLite.',
          type: 'array',
          required: 'optional',
          calculated: true

attribute 'owncloud/packages/mysql',
          display_name: 'owncloud packages mysql',
          description: 'ownCloud package names array for MySQL.',
          type: 'array',
          required: 'optional',
          calculated: true

attribute 'owncloud/packages/postgresql',
          display_name: 'owncloud packages postgresql',
          description: 'ownCloud package names array for PostgreSQL.',
          type: 'array',
          required: 'optional',
          calculated: true
